### PR TITLE
Ceara/aitt 408 accuracy threshold

### PR DESCRIPTION
**Pull request by cearachew**
Jira Ticket: https://codedotorg.atlassian.net/browse/AITT-408

Adds an option to `rubric_tester` to compare overall accuracy and accuracy by key concept in each lesson to thresholds in the `accuracy_thresholds.json` file. This also adds a pytest test that can be run (does not run by default) with the flag `--accuracy` to pytest or with the convenience script `bin/test_accuracy.sh`. This test passes if the accuracy meets or exceeds the thresholds. If any lesson has overall accuracy or accuracy for a key concept less than the threshold, the test fails and gives output listing the failed accuracy checks. 

This test does not run on Docker, only locally. The accuracy threshold is set to 0.7 for all lessons and key concepts for now.

Follow-up: Ideally would like to move away from fetching lesson files from Google Drive, since it's easy for the Google Drive files to be in a bad state from someone working on them. I think pulling them from S3 for regression testing would be better. There's another ticket for getting config params from S3 here, and I think it would make sense to do that follow up with that: https://codedotorg.atlassian.net/browse/AITT-406

Failed output looks like: 
```
✗ bin/test_accuracy.sh          
Running: `coverage run -m pytest --accuracy  && coverage report -m`
========================================= test session starts ==========================================
platform darwin -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/cearachew/code/aiproxy
plugins: mock-3.12.0, anyio-3.7.1, requests-mock-1.11.0
collected 109 items                                                                                    

tests/accuracy/test_accuracy.py F                                                                [  0%]
tests/routes/test_assessment_routes.py ........................                                  [ 22%]
tests/routes/test_openai_routes.py .....                                                         [ 27%]
tests/routes/test_simple_routes.py .....                                                         [ 32%]
tests/unit/assessment/test_assessment.py .....                                                   [ 36%]
tests/unit/assessment/test_label.py ......................................                       [ 71%]
tests/unit/assessment/test_report.py ................                                            [ 86%]
tests/unit/assessment/test_rubric_tester.py ...............                                      [100%]

=============================================== FAILURES ===============================================
______________________________________ TestAccuracy.test_accuracy ______________________________________

self = <test_accuracy.TestAccuracy object at 0x111385550>

    def test_accuracy(self):
        assert "OPENAI_API_KEY" in os.environ
        with mock.patch('sys.argv', ['rubric_tester.py', '-a']):
            ret = main()
>       assert ret == True
E       assert False == True

tests/accuracy/test_accuracy.py:19: AssertionError
----------------------------------------- Captured stdout call -----------------------------------------
Namespace(lesson_names=dict_keys(['U3-2022-L17', 'U3-2022-L20', 'New-U3-2022-L10', 'New-U3-2022-L13']), output_filename='report.html', use_cached=False, llm_model='gpt-4', num_responses=1, num_passing_labels=None, max_num_students=100, student_ids=None, temperature=0.0, download=False, accuracy=True, passing_labels=None)
FAIL
------------------------------------------ Captured log call -------------------------------------------
ERROR    root:rubric_tester.py:310 The following thresholds were not met:
{ 'New-U3-2022-L10': { 'key_concepts': { 'Position - Elements and the Coordinate System': { 'accuracy_score': 0.5555555555555556,
                                                                                            'threshold': 0.7}}},
  'New-U3-2022-L13': { 'key_concepts': { 'Position and Movement': { 'accuracy_score': 0.55,
                                                                    'threshold': 0.7}}},
  'U3-2022-L17': { 'key_concepts': { 'Position and Movement': { 'accuracy_score': 0.5652173913043478,
                                                                'threshold': 0.7}}},
  'U3-2022-L20': { 'key_concepts': { 'Algorithms and Control Structures': { 'accuracy_score': 0.2,
                                                                            'threshold': 0.7},
                                     'Program Development 2': { 'accuracy_score': 0.5,
                                                                'threshold': 0.7},
                                     'Variables': { 'accuracy_score': 0.65,
                                                    'threshold': 0.7}},
                   'overall': {'accuracy_score': 0.45, 'threshold': 0.7}}}
======================================= short test summary info ========================================
```

### CODE REVIEW

File: lib/assessment/config.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 SUPPORTED_MODELS = ['gpt-4', 'gpt-4-0314', 'gpt-4-32k', 'gpt-4-32k-0314']
 LESSONS = {
-    "U3-2022-L10" : "1ROCbvHb3yWGVoQqzKAjwdaF0dSRPUjy_",
-    "U3-2022-L13" : "1kGHeY5LRpFJ9xVRoBEWbyOJyKm4wClqw",
+    # "U3-2022-L10" : "1ROCbvHb3yWGVoQqzKAjwdaF0dSRPUjy_",
+    # "U3-2022-L13" : "1kGHeY5LRpFJ9xVRoBEWbyOJyKm4wClqw",
     "U3-2022-L17" : "1WirJLIFgo-anxAz-kZXDVQ2Tl_8OuX22",
     "U3-2022-L20" : "115BHvZ1kJC2xhUSOBkLiE8DC1YgcjyRd",
-    "U3-2022-L23" : "12OJex4l9OhWrnbLenpvZAibtfiFWWdzx",
+    # "U3-2022-L23" : "12OJex4l9OhWrnbLenpvZAibtfiFWWdzx",
```
Comment (davidsbailey) [inline lib/assessment/config.py:8]: did you mean to leave these commented out?

File: tests/accuracy/test_accuracy.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+@accuracy
+@pytest.mark.accuracy_setup
+class TestAccuracy:
+    def test_accuracy(self):
+        assert "OPENAI_API_KEY" in os.environ
+        with mock.patch('sys.argv', ['rubric_tester.py', '-a']):
+            ret = main()
+        assert ret == True
+
```
Comment (davidsbailey) [inline tests/accuracy/test_accuracy.py:20]: Please configure your editor to always end each file with a trailing newline. the ⛔ symbol shown here by github indicates the newline is missing.

File: TESTING.md [hunk 3]
```diff
@@ -26,6 +27,21 @@ just run `pytest` within a running container's shell session by using the
 PYTHONPATH=/app pytest
 ```
 
+## Accuracy Tests
+
+**Running the Accuracy test hits the OpenAI endpoint and is expensive! Only run this test infrequently**
+
+To run the accuracy threshold test, follow directions in `README.md` to set up your local
+environment for running the Rubric Tester. You can then run `.bin/test_accuracy.sh` to run
```
Comment (davidsbailey) [inline TESTING.md:35]: nit: `./bin`

**Comment by davidsbailey** [comment]
In case we can't resolve my setup problem, I'm hoping you can share what steps you ran to verify that this is working. do you get a passing result when you run `./bin/test_accuracy.sh` against what's in google drive?

File: lib/assessment/config.py (hunk 1 above)
Comment (cearachew) [reply lib/assessment/config.py:8]: I did, these files in google drive were not in a good state. This, along with your issues pulling down from google drive, are why I'd prefer to put the test files and actual results in s3 so we pull a more "final" version of the files with more explicit authentication than whatever gdown is doing

**Comment by cearachew** [comment]
> In case we can't resolve my setup problem, I'm hoping you can share what steps you ran to verify that this is working. do you get a passing result when you run `./bin/test_accuracy.sh` against what's in google drive?

I ran `./bin/test_accuracy.sh` and get the result that's in the PR comment. It doesn't exceed all the thresholds, which are set to 0.7 for everything, but it does run. If you want to verify it without the problem with downloading from google drive using gdown, you could manually copy the files from google drive to lesson dirs in `lesson_data/` and rerun.

**Comment by davidsbailey** [comment]
> > In case we can't resolve my setup problem, I'm hoping you can share what steps you ran to verify that this is working. do you get a passing result when you run `./bin/test_accuracy.sh` against what's in google drive?
> 
> I ran `./bin/test_accuracy.sh` and get the result that's in the PR comment. It doesn't exceed all the thresholds, which are set to 0.7 for everything, but it does run. If you want to verify it without the problem with downloading from google drive using gdown, you could manually copy the files from google drive to lesson dirs in `lesson_data/` and rerun.

Thank you for clarifying, I wasn't sure if you had done something special to force it into the failure case. I think it is ok for the test to be failing for now. Let's get it pointing at S3 before we debug the cause of the low accuracy scores.

File: lib/assessment/config.py (hunk 1 above)
Comment (davidsbailey) [reply lib/assessment/config.py:8]: this is fine, thank you for clarifying. it was an honest question :-) I think it is ok to switch to S3 in the next PR.

**Review by davidsbailey — APPROVED** [review]

**Comment by davidsbailey** [comment]
@cearachew just checking in on whether anything is blocking you from merging this PR? If it is ready, I'd love to be able to start using it soon.

**Comment by cearachew** [comment]
@davidsbailey Oops, nothing blocking! I just thought I had already merged it 🤦‍♀️

**Pull request merged**